contracts/test/StakeTable.t.sol

```diff
@@ -0,0 +1,225 @@
+        assertEq(nativeAmount, depositAmount);
+        assertEq(restakedAmount, 0);
+    }
+
+    function testInvalidBLSSig() external {
+        (BN254.G2Point memory blsVK, EdOnBN254.EdOnBN254Point memory schnorrVK,) =
+            computeGroupElements(msg.sender);
+
+        // Failed signature verification
+        BN254.G1Point memory badSig = BN254.P1();
+        vm.expectRevert(BLSSig.BLSSigVerificationFailed.selector);
+        stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Native, badSig, validUntilEpoch
+        );
+    }
+
+    function testRestakingNotImplemented() external {
+        (
+            BN254.G2Point memory blsVK,
+            EdOnBN254.EdOnBN254Point memory schnorrVK,
+            BN254.G1Point memory sig
+        ) = computeGroupElements(msg.sender);
+
+        // Throw "Restaking not implemented" error
+        vm.expectRevert(S.RestakingNotImplemented.selector);
+        stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Restake, sig, validUntilEpoch
+        );
+    }
+
+    function testInvalidNextRegistrationEpoch() external {
+        (
+            BN254.G2Point memory blsVK,
+            EdOnBN254.EdOnBN254Point memory schnorrVK,
+            BN254.G1Point memory sig
+        ) = computeGroupElements(msg.sender);
+
+        // Invalid next registration epoch
+        vm.prank(msg.sender);
+        vm.expectRevert(abi.encodeWithSelector(S.InvalidNextRegistrationEpoch.selector, 1, 0));
+        stakeTable.register(blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Native, sig, 0);
+    }
+
+    function testNodeAlreadyRegistered() external {
+        (
+            BN254.G2Point memory blsVK,
+            EdOnBN254.EdOnBN254Point memory schnorrVK,
+            BN254.G1Point memory sig
+        ) = computeGroupElements(msg.sender);
+
+        // Prepare for the token transfer
+        vm.prank(msg.sender);
+        token.approve(address(stakeTable), depositAmount);
+
+        // Successful call to register
+        vm.prank(msg.sender);
+        stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Native, sig, validUntilEpoch
+        );
+
+        // The node is already registered
+        vm.prank(msg.sender);
+        vm.expectRevert(S.NodeAlreadyRegistered.selector);
+        stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Native, sig, validUntilEpoch
+        );
+    }
+
+    function testTransferFailed() external {
+        (
+            BN254.G2Point memory blsVK,
+            EdOnBN254.EdOnBN254Point memory schnorrVK,
+            BN254.G1Point memory sig
+        ) = computeGroupElements(msg.sender);
+
+        vm.prank(msg.sender);
+        vm.expectRevert("TRANSFER_FROM_FAILED");
+        stakeTable.register(
+            blsVK, schnorrVK, depositAmount, IStakeTable.StakeType.Native, sig, validUntilEpoch
+        );
```

Comment:
we forget to check: if the balance is correct if transfer failed.

in fact, we should add "[invariant test](https://book.getfoundry.sh/forge/invariant-testing)" w.r.t. the total balance of the StakeTable contract and of the users. 
^^ for this, we can add slightly later, not in this PR.

Reply:
Good idea, ticket #737 updated.